Full Store Editing: Technical documentation (#5939)
woocommerce/woocommerce-gutenberg-products-block: docs/templates/block-template-controller.md

# BlockTemplateController
[Source file](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/BlockTemplatesController.php)

> The BlockTemplateController class contains all the business logic which loads the templates into the Site Editor or on the front-end through various hooks available in WordPress & WooCommerce core. Without documenting every method individually, I will look to provide some insight into key functionality. 

Reviewer: ```suggestion
The `BlockTemplateController` class contains all the business logic which loads the templates into the Site Editor or on the front-end through various hooks available in WordPress and WooCommerce core. Without documenting every method individually, I will look to provide some insight into (its?) key functionality. 
```

## Table of Contents

* [Overview](#overview)
* [add_block_templates( $query_result, $query, $template_type )](#add_block_templates-query_result-query-template_type-)

> * [get_block_file_template( $template, $id, $template_type )](#get_block_file_template-template-id-template_type-)

Reviewer: Just double checking, is the ending `-` correct as an anchor?

Author: Yep, this is correct.

**Typically executed when:**
* Loading the "All Templates" view in the Site Editor
* Loading one of the templates on the frontend where the query would build a list of relevant templates based on a hierarchy (for example, the product page hierarchy could be an array containing  `single-product-[product-name].html`, `single-product.html`, `single.html`).

**This method is responsible for:**

* Giving our templates a user-friendly title (e.g. turning "single-product" into "Product Page").
* It collects all the WooCommerce templates from both the filesystem and the database (customized templates are stored in the database as posts) and adds them to the returned list.
* In the event the theme has a `archive-product.html` template file, but not category/tag template files, it is eligible to use the `archive-product.html` file in their place. So we trick Gutenberg in thinking some templates (e.g. category/tag) have a theme file available if it is using the `archive-product.html` template, even though _technically_ the theme does not have a specific file for them.

> * If the query is targeting templates for a specific post type (e.g. on the Edit Post view), we do not add ineligible WooCommerce templates to the returned list.

Reviewer: I'm reading this docs without looking at the source code for added realism 😁 , and I am not sure I understand what this means. Is there any way we can make this clearer?

Author: Hmm yes, let me work on rewording this.

Author: Hopefully this is an improvement https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5939/commits/c0bf7a0cd668d45986dc11310f6d1f8a3ce1a235

### Return value

This method will return an array of `WP_Block_Template` values

## get_block_file_template( $template, $id, $template_type )

> This method is applied to the filter `pre_get_block_file_template` inside the WordPress core function `get_block_file_template` (not to be confused with this method from the BlockTemplateController, of which has the same name). 

Reviewer: ```suggestion
This method is applied to the filter `pre_get_block_file_template` inside the WordPress core function `get_block_file_template` (not to be confused with this method from the `BlockTemplateController`, of which has the same name). 
```

The order of execution is as follows:

1. `get_block_template()` from WordPress core will execute, and attempt to retrieve a customized version of the template from the database.

> 2. If it fails to retrieve one, it will execute the `get_block_file_template()` from WordPress core which will apply the filter `pre_get_block_file_template` which is what we hook into to return our template file, which will trigger an early return to prevent WordPress from continuing its query.

Reviewer: Any way we can contain the “whiches”? So many relative clauses are a bit hard to parse, especially as a non native speaker, I suppose.

One way I can think of would be to just stop the sentence and resume another one, something like:

_If it fails to retrieve one, it will execute the `get_block_file_template()` from WordPress core which will apply the filter `pre_get_block_file_template`. This is were we hook into to return our template file, and trigger an early return to prevent WordPress from continuing its query_

Author: Nice! Reads much better

During step 2 it's important we hook into the `pre_get_block_file_template` because if we don't, the function will check if the first part of the template ID (e.g. `woocommerce/woocommerce`) is the same as the current themes ID (e.g. `twentytwentytwo`), which will resolve `false` and return `null` instead of the expected `WP_Block_Template` object.

**Typically executed when:**
* A user clears the customizations of a WooCommerce template.

**This method is responsible for:**
* Loading the template files from the filesystem, and building a `WP_Block_Template` version of it.

### Return value

This method will return `WP_Block_Template` or `null`

## render_block_template()

This method is applied to the filter `template_redirect` and executed before WordPress determines which template to load.

This allows us to hook into WooCommerce core through the filter `woocommerce_has_block_template` where we can determine if a specific block template exists and should be loaded.

**Typically executed when:**
* A user loads a page on the front-end.

**This method is responsible for:**
* Determining if the current page has an appropriate WooCommerce block template available to render.

> * Checking if the currently loaded page is from WooCommerce. It then checks if the theme has an appropriate template to use, if it does not then it finally checks if WooCommerce has a default block template available. If so, we override the value through `woocommerce_has_block_template` to resolve true.

Reviewer: ```suggestion
* Checking if the currently loaded page is from WooCommerce. It then checks if the theme has an appropriate template to use: if it does not, then it finally checks if WooCommerce has a default block template available. If so, we override the value through `woocommerce_has_block_template` to resolve `true`.
```

Within each method section, I will explain in what scenarios they are executed. 

* filter: `get_block_templates` with `add_block_templates`.
* filter: `pre_get_block_file_template` with `get_block_file_template`.
* action: `template_redirect` with `render_block_template`.

## add_block_templates( $query_result, $query, $template_type )

This method is applied to the filter `get_block_templates`, which is executed before returning a unified list of template objects based on a query.

**Typically executed when:**
* Loading the "All Templates" view in the Site Editor
* Loading one of the templates on the front-end where the query would build a list of relevant templates based on a hierarchy (for example, the product page hierarchy could be an array containing  `single-product-[product-name].html`, `single-product.html`, `single.html`).
* Loading the "Edit Product" view.

**This method is responsible for:**

* Giving our templates a user-friendly title (e.g. turning "single-product" into "Product Page").

> * It collects all the WooCommerce templates from both the filesystem and the database (customized templates are stored in the database as posts) and adds them to the returned list.

Reviewer 2: ```suggestion
* Collecting all the WooCommerce templates from both the filesystem and the database (customized templates are stored in the database as posts) and adding them to the returned list.
```

* In the event the theme has a `archive-product.html` template file, but not category/tag template files, it is eligible to use the `archive-product.html` file in their place. So we trick Gutenberg in thinking some templates (e.g. category/tag) have a theme file available if it is using the `archive-product.html` template, even though _technically_ the theme does not have a specific file for them.
* Ensuring we do not add irrelevant WooCommerce templates in the returned list. For example, if  `$query['post_type']` has a value (e.g. `product`) this means the query is requesting templates related to that specific post type, so we filter out any irrelevant ones. This _could_ be used to show/hide templates from the template dropdown on the "Edit Product" screen in WP Admin.

### Return value

> This method will return an array of `WP_Block_Template` values

Reviewer 2: ```suggestion
This method will return an array of `WP_Block_Template` values.
```

## get_block_file_template( $template, $id, $template_type )

> This method is applied to the filter `pre_get_block_file_template` inside the WordPress core function `get_block_file_template` (not to be confused with this method from the `BlockTemplateController` class, of which has the same name). 

Reviewer 2: ```suggestion
This method is applied to the filter `pre_get_block_file_template` inside the WordPress core function `get_block_file_template` (not to be confused with this method from the `BlockTemplateController` class, which has the same name). 
```

The order of execution is as follows:

1. `get_block_template()` from WordPress core will execute, and attempt to retrieve a customized version of the template from the database.

> 2. If it fails to retrieve one, it will execute the `get_block_file_template()` from WordPress core which will apply the filter `pre_get_block_file_template`. This is where we hook into to return our template file, and trigger an early return to prevent WordPress from continuing its query.

Reviewer 2: ```suggestion
2. If it fails to retrieve one, it will execute the `get_block_file_template()` function from WordPress core which will apply the filter `pre_get_block_file_template`. This is where we hook in to return our template file, and trigger an early return to prevent WordPress from continuing its query.
```

> During step 2 it's important we hook into the `pre_get_block_file_template` because if we don't, the function will check if the first part of the template ID (e.g. `woocommerce/woocommerce`) is the same as the current themes ID (e.g. `twentytwentytwo`), which will resolve `false` and return `null` instead of the expected `WP_Block_Template` object.

Reviewer 2: ```suggestion
During step 2 it's important we hook into the `pre_get_block_file_template`. If we don't, the function will check if the first part of the template ID (e.g. `woocommerce/woocommerce`) is the same as the current themes ID (e.g. `twentytwentytwo`), which will resolve `false` and return `null` instead of the expected `WP_Block_Template` object.
```

**Typically executed when:**
* A user clears the customizations of a WooCommerce template.

**This method is responsible for:**
* Loading the template files from the filesystem, and building a `WP_Block_Template` version of it.

### Return value

> This method will return `WP_Block_Template` or `null`

Reviewer 2: ```suggestion
This method will return `WP_Block_Template` or `null`.
```
